src/mds/Server.cc

```diff
@@ -939,11 +939,25 @@ void Server::find_idle_sessions()
     std::vector<Session*> new_stale;
 
     for (auto session : *(sessions_p1->second)) {
+      client_t client = session->get_client();
       auto last_cap_renew_span = std::chrono::duration<double>(now - session->last_cap_renew).count();
+
       if (last_cap_renew_span < cutoff) {
 	dout(20) << "laggiest active session is " << session->info.inst
 		 << " and renewed caps recently (" << last_cap_renew_span << "s ago)" << dendl;
 	break;
+      } else if (last_cap_renew_span < mds->mdsmap->get_session_autoclose()
+		 && !(mds->locker->is_revoking_any_caps_from(client) ||
+		 session->is_any_flush_waiter())) {
+	dout(20) << "deferring marking session " << client.v << " stale "
+		    "since it holds no caps" << dendl;
+	continue;
+      } else {
+	dout(20) << "evicting session " << client.v << " since autoclose "
+		    "has arrived" << dendl;
+	// evict session without marking it stale
+	to_evict.push_back(session);
+	continue;
       }
 
       if (session->last_seen > session->last_cap_renew) {
```

Comment:
your code should be added after "if (session->last_seen > session->last_cap_renew) "

Reply:
Done.